Reduce lock contention for MergeTree tables (by renaming parts without holding lock)

azat opened the pull request:
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
This change was reverted


Reduce lock contention for MergeTree tables (by renaming parts without holding lock)

Under heavy load, or not so heavy but with fsync_part_directory=1, time that renameTo() holds DataPartsLock will be increased, and this will affect almost every operation with this table.

On one of production clusters I saw ~60 seconds with fsync_part_directory=1.

Move the renameTo() out from the critical section.

azat committed:
Reduce lock contention for MergeTree tables (by renaming parts without holding lock)

Under heavy load, or not so heavy but with fsync_part_directory=1,
time that renameTo() holds DataPartsLock will be increased, and this
will affect almost every operation with this table.

On one of production clusters I saw ~60 seconds with
fsync_part_directory=1.

Move the renameTo() out from the critical section.

v2: instead of using DataPartsLock.lock.lock()/unlock() move the renameTo() into MergeTreeData::Transaction::commit()
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Remove unused locked_parts from MergeTreeData::Transaction

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #61973 from azat/ClickHouse:mt/rename-without-lock -->
*This is an automated comment for commit 3c10ef32a8420f0fa939809a72ff5ae1be7df28d with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61973/3c10ef32a8420f0fa939809a72ff5ae1be7df28d/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

azat committed:
Fix replacing parts with empty

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
Likely related test failures:

>Stateless tests (asan) [2/4] — Tests are not finished, fail: 1, passed: 383, skipped: 6
>Stateless tests (tsan) [5/5] — Tests are not finished, fail: 3, passed: 302, skipped: 2
>Stateless tests (tsan, s3 storage) [5/5] — Tests are not finished, fail: 2, passed: 298, skipped: 4

```
2024.03.28 06:23:18.782173 [ 937 ] {} <Fatal> : Logical error: 'Trying to remove detached part detached/attaching_all_21_21_0 with path store/b9e/b9ea0283-b7ac-48cd-ab64-54e326a59f96/ in remove function. It shouldn't happen'.
```

Unrelated test failures:

>Stress test (asan) — Found signal in gdb.log

- https://github.com/ClickHouse/ClickHouse/issues/62062

>Stateless tests (asan) [1/4] — fail: 1, passed: 1579, skipped: 17
>Stateless tests (debug, s3 storage) [5/6] — fail: 1, passed: 1047, skipped: 27
>Stateless tests (msan) [5/6] — fail: 1, passed: 1066, skipped: 8

- https://github.com/ClickHouse/ClickHouse/issues/56052

>Stateless tests (coverage) [2/6] — Some queries hung, fail: 1, passed: 1021, skipped: 3

- https://github.com/ClickHouse/ClickHouse/issues/54748

>Stateless tests (release) — fail: 1, passed: 6280, skipped: 9

- https://github.com/ClickHouse/ClickHouse/issues/62063

azat committed:
Use renameParts() explicitly to avoid leaving parts in detached

Since there is an assertion that does not allows to remove detached
parts during cleanup, which sounds good in general, but breaks this new
code.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Avoid race between cleanup thread and renameMergedTemporaryPart()

The problem was that with this patch set renameMergedTemporaryPart() is
called without temporary_directory_lock holded (in MergeTask), since it
is reseted just before calling renameMergedTemporaryPart(), and this can
be seen in logs:

    2024.03.29 19:56:42.126919 [ 1341 ] {ea7a3fd2-cf47-4ec7-91a5-51c69fba1b95::-8_0_138_2_2} <Trace> test_btnct5cr.alter_table_0 (ea7a3fd2-cf47-4ec7-91a5-51c69fba1b95) (MergerMutator): Merged 50 parts: [-8_0_0_0_2, -8_138_138_0] -> -8_0_138_2_2
    2024.03.29 19:56:42.127034 [ 1341 ] {ea7a3fd2-cf47-4ec7-91a5-51c69fba1b95::-8_0_138_2_2} <Debug> test_btnct5cr.alter_table_0 (ea7a3fd2-cf47-4ec7-91a5-51c69fba1b95): Committing part -8_0_138_2_2 to zookeeper
    2024.03.29 19:56:42.128462 [ 884 ] {} <Warning> test_btnct5cr.alter_table_0 (ea7a3fd2-cf47-4ec7-91a5-51c69fba1b95): Removing temporary directory /var/lib/clickhouse/store/ea7/ea7a3fd2-cf47-4ec7-91a5-51c69fba1b95/tmp_merge_-8_0_138_2_2/
    2024.03.29 19:56:42.128647 [ 1341 ] {ea7a3fd2-cf47-4ec7-91a5-51c69fba1b95::-8_0_138_2_2} <Debug> test_btnct5cr.alter_table_0 (ea7a3fd2-cf47-4ec7-91a5-51c69fba1b95): Part -8_0_138_2_2 committed to zookeeper
    ...
    2024.03.29 19:56:54.586084 [ 57841 ] {bf240267-0620-4294-afc1-479c58e6be89} <Error> executeQuery: std::exception. Code: 1001, type: std::__1::__fs::filesystem::filesystem_error, e.what() = filesystem error: in file_size: No such file or directory ["/var/lib/clickhouse/store/ea7/ea7a3fd2-cf47-4ec7-91a5-51c69fba1b95/-8_0_138_2_2/data.cmrk3"]

This should fix failures of 00993_system_parts_race_condition_drop_zookeeper in [1].

  [1]: https://s3.amazonaws.com/clickhouse-test-reports/61973/f6f826c85dd5b7bb8db16286fd10dcf441a440f7/stateless_tests__coverage__[4_6].html

Though now it looks hackish...

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
Related:

>Stateless tests (coverage) [4/6] — fail: 1, passed: 1092, skipped: 2

- `00993_system_parts_race_condition_drop_zookeeper`

```
2024-03-29 19:57:24 [d922bb02cd61] 2024.03.29 00:56:54.586084 [ 57841 ] {bf240267-0620-4294-afc1-479c58e6be89} <Error> executeQuery: std::exception. Code: 1001, type: std::__1::__fs::filesystem::filesystem_error, e.what() = filesystem error: in file_size: No such file or directory ["/var/lib/clickhouse/store/ea7/ea7a3fd2-cf47-4ec7-91a5-51c69fba1b95/-8_0_138_2_2/data.cmrk3"]
```

Unrelated:

>Stateless tests (tsan) [3/5] — Tests are not finished, fail: 2, passed: 483, skipped: 6
>Stateless tests (msan) [3/6] — Tests are not finished, fail: 1, passed: 367, skipped: 7
>Stateless tests (debug) [3/5] — Tests are not finished, fail: 5, passed: 484, skipped: 7
>Stateless tests (asan) [1/4] — Tests are not finished, fail: 2, passed: 569, skipped: 8

#62062

azat commented:
Still some issues:

>[Stateless tests (debug) [5/5] — Tests are not finished, fail: 9, passed: 361, skipped: 2](https://s3.amazonaws.com/clickhouse-test-reports/61973/adcf7d5b2ec4d904db51be5d545aa5d497a04c4e/stateless_tests__debug__[5_5].html)

```
2024.04.03 02:03:25.724784 [ 1120 ] {} <Fatal> : Logical error: 'Trying to remove detached part detached/all_200_200_0 with path store/6f2/6f24250b-14de-45e1-8b44-f0f25710765d/ in remove function. It shouldn't happen'.
```

azat committed:
Add reason into "Part {} is broken and need manual correction" message

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Require explicit rename of parts in transaction

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Do not remove detached parts in Transaction::rollback

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix possible assertion when size of precommitted_parts <= precommitted_parts_need_rename

CI founds [1]:

    Logical error: 'precommitted_parts.size() >= precommitted_parts_need_rename.size()'

  [1]: https://s3.amazonaws.com/clickhouse-test-reports/61973/5c1e6a3e956917bdbb7eaa467934e5b75f17a923/stateless_tests__tsan__s3_storage__[5_5].html

The problem is that after precommitted_parts cleaned from detached parts
it may be less then precommitted_parts_need_rename, so to avoid this,
let's just copy it to a new container.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>Stateless tests (release, old analyzer, s3, DatabaseReplicated) [4/4] — fail: 1, passed: 1539, skipped: 27

- `03093_bug37909_query_does_not_finish` - 4f6b6e30e101

azat commented:
Can someone take a look at this one?

azat commented:
Kind ping

alexey-milovidov commented:
@azat, need to fix all the tests.

azat commented:
@alexey-milovidov tests failures does not looks related, but it is quite old, so I will do a rebase at least.

azat commented:
>ClickHouse special build check — 12/13 artifact groups are OK

- tidy - https://github.com/ClickHouse/ClickHouse/pull/64349#issuecomment-2148883998

azat commented:
>Stateless tests (release) — fail: 1, passed: 6629, skipped: 13

- `02956_rocksdb_bulk_sink` - https://github.com/ClickHouse/ClickHouse/pull/64856

azat commented:
>Stateless tests (debug, s3 storage) [3/6] fail: 1, passed: 1071, skipped: 24

- https://github.com/ClickHouse/ClickHouse/pull/64887

>Stateless tests (tsan, s3 storage) [1/5] fail: 1, passed: 1329, skipped: 24

- https://github.com/ClickHouse/ClickHouse/pull/64888

alexey-milovidov commented:
@azat, please resubmit.
